### PR TITLE
Sampling tests2

### PR DESCRIPTION
**Pull request by rleonid**

### CODE REVIEW

File: src/lib/descriptive.ml [hunk 1]
```diff
@@ -41,26 +53,35 @@ let skew arr =
...
-  float (Array.length arr)
+  let n = float (Array.length arr) in
+  let v = unbiased_var arr in
+  sqrt (2.0 /. (n -. 1.0)) *. v
 
 let skew_standard_error arr =
-  sqrt ( 6.0 /. (float (Array.length arr)))
+  (* Older: sqrt ( 6.0 /. (float (Array.length arr))) *)
+  let n = float (Array.length arr) in
+  sqrt ((6.0*.n*.(n-.1.0)) /. ((n-.2.0)*.(n+.1.0)*.(n+.3.0)))
```
Comment (smondet) [inline src/lib/descriptive.ml]: Styling: missing spaces.

(A whole overloading module, `Float.(6. * n * (n - 1.) .... )` would be nice also for readability).

File: src/lib/descriptive.ml [hunk 2]
```diff
@@ -71,66 +92,53 @@ let skew_statistic arr =
 let kurtosis_statistic arr =
   (unbiased_kurtosis arr) /. (kurtosis_standard_error arr)
 
+type skew_classification =
+  [ `Negative | `SlightNegative | `Normal | `SlightPositive | `Positive ]
```
Comment (smondet) [inline src/lib/descriptive.ml]: ``Slight_negative`

File: src/lib/descriptive.mli [hunk 3]
```diff
@@ -1,35 +1,146 @@
...
+(** Spread. *)
+
+(** [var data] returns the sample variance of [data]. *)
 val var : float array -> float
+
+(** [unbiased_var] returns the unbiased variance of [data] via Bessel's
+    correction; ie. divinding by [n - 1]. *)
 val unbiased_var : float array -> float
+
+(** [population_var mean data] compute the variance of data when you know the
```
Comment (smondet) [inline src/lib/descriptive.mli]: compute**S**

File: src/lib/descriptive.mli [hunk 4]
```diff
@@ -1,35 +1,146 @@
...
+  population [mean]. *)
+val population_var : float -> float array -> float
+
+(* Between two random variables.*)
+
+(** [covariance x y] compute the sample covariance of [x] and [y] where higher
```
Comment (smondet) [inline src/lib/descriptive.mli]: compute**S**

File: src/lib/descriptive.mli [hunk 5]
```diff
@@ -1,35 +1,146 @@
...
+(** [autocorrelation lag data] computes the correlation of [data] with itself
+  offset by [lag].  *)
+val autocorrelation : int -> float array -> float
+
+(* Higher moments.*)
+
+(** [moment k data] computes the [k]th sample central moment of [data]. *)
 val moment : int -> float array -> float
+
+(** [skew data] compute the sample skew of [data] (Fisher-Pearson's moment
```
Comment (smondet) [inline src/lib/descriptive.mli]: computes

File: src/lib/descriptive.mli [hunk 6]
```diff
@@ -1,35 +1,146 @@
...
+(** [classify_skew data] provides a terse description of the skewness of [data].
+
+    Please note, even though a (polymorphic) variant is returned, caution must
+    be taken when creating algorithms based upon these classification. These
+    methods are brittle in the face of non-normal data and small sample size,
+    although effort has been made to choose good practices.  *)
+val classify_skew : float array -> skew_classification
+
+type kurtosis_classification =
+  [ `Skinny | `SlightlySkinny | `Fat | `SlightlyFat | `Normal ]
```
Comment (smondet) [inline src/lib/descriptive.mli]: `No_camel_case`

File: src/lib/functions.mli [hunk 7]
```diff
@@ -16,4 +16,9 @@ val erfc : float -> float
 val chi_square_less : float -> int -> float
 val chi_square_greater : float -> int -> float
 val t_lookup : float -> int -> float
-val softmax : ?temp:float -> float array -> int -> float
+
+(** [softmax ?temp weights] transform [weights] into softmax weights dependent
```
Comment (smondet) [inline src/lib/functions.mli]: transform**S**

File: src/lib/functions.mli [hunk 8]
```diff
@@ -16,4 +16,9 @@ val erfc : float -> float
...
+    on [temp]erature.
+
+    @raise Invalid_argument if [weights] is empty or [temp = 0]. *)
+val softmax : ?temp:float -> float array -> float array
```
Comment (smondet) [inline src/lib/functions.mli]: I'd put `?temperature`, `?temp` out of context would be more “temporary,” and that kind of abbreviation is not worth it.

File: src/lib/svd.mli [hunk 9]
```diff
@@ -25,5 +25,9 @@ val covariance_matrix : ?lambda:float -> t -> mat
     vector, for computing the optimal regularization parameter. *)
 val looe : t -> vec -> (float -> vec)
 
+(** [h t] compute the Hat matrix of the original matrix passed to Svd. *)
```
Comment (smondet) [inline src/lib/svd.mli]: compute**S**

File: src/lib/svd.mli [hunk 10]
```diff
@@ -25,5 +25,9 @@ val covariance_matrix : ?lambda:float -> t -> mat
...
 val h : t -> mat
+
+(** [h_diag t] compute the diagonal of the Hat matrix of the original matrix
```
Comment (smondet) [inline src/lib/svd.mli]: idem

**Comment by smondet** [comment]
LGTM modulo some minor comments above :)

File: src/lib/descriptive.ml (hunk 1 above)
Comment (rleonid) [inline src/lib/descriptive.ml]: Good point, I like the adding of adding an overloading module.

File: src/lib/descriptive.mli (hunk 3 above)
Comment (rleonid) [inline src/lib/descriptive.mli]: Thank you.

File: src/lib/descriptive.mli (hunk 6 above)
Comment (rleonid) [inline src/lib/descriptive.mli]: :+1:

File: src/lib/functions.mli (hunk 8 above)
Comment (rleonid) [inline src/lib/functions.mli]: Good point.

**Pull request merged**